### PR TITLE
Added node tuning operator content

### PR DESCRIPTION
**Pull request by ahardin-rh**
WIP Preview Build: http://file.rdu.redhat.com/~ahardin/Feb182019/node-tuning-operator/scalability_and_performance/using-node-tuning-operator.html

Dev card: https://jira.coreos.com/browse/ATS-185

Docs card: https://jira.coreos.com/browse/OSDOCS-45

### CODE REVIEW

**Comment by ahardin-rh** [comment]
@sjug @jmencak FYI, just stubbed out this section so that we could fill in the content.

**Comment by jmencak** [comment]
Good start, thank you Ashley.  I believe that a lot of information could be lifted from https://github.com/openshift/cluster-node-tuning-operator/blob/master/README.md

Things I'd add could be a mention of the default profiles set on a cluster, i.e. https://github.com/openshift/cluster-node-tuning-operator/blob/master/assets/tuned/07-cr-tuned.yaml

Would you be able to put something together for me to review?

**Comment by ahardin-rh** [comment]
@jmencak Thanks and sorry for the delay! I made some additions. 

Preview build:
http://file.rdu.redhat.com/~ahardin/11142018/node-tuning-operator/scalability_and_performance/using-node-tuning-operator.html

File: modules/deploying-node-tuning-operator.adoc
Comment (ahardin-rh) [inline modules/deploying-node-tuning-operator.adoc]: I feel like we can tighten this procedure up.

Comment (jmencak) [reply modules/deploying-node-tuning-operator.adoc]: What do you mean by tighten up?

**Comment by jmencak** [comment]
Thanks Ashley, for OCP, the operator will be automatically deployed, so the deployment section will change for sure.  At the moment it is deployed by automatically by CVO, but might change.

File: modules/deploying-node-tuning-operator.adoc
Comment (ahardin-rh) [reply modules/deploying-node-tuning-operator.adoc]: @jmencak We list these steps, but they are not very explicit:

1. Apply the Operator’s manifest to the desired namespace in the cluster.
2. Create an application instance.
3. Update an application.

Are there commands that need to run to accomplish these steps? How would the user do these tasks?

Comment (jmencak) [reply modules/deploying-node-tuning-operator.adoc]: I agree, but this hasn't been finalized, that's why it is as it is.  The user shouldn't need to deploy the operator if it is installed by CVO.  The user should only be concerned by creating a CR if he/she needs it.

**Comment by ahardin-rh** [comment]
@jmencak I'll hold off on merging this until those details are sorted out. 
@jeremyeder FYI

**Comment by ahardin-rh** [comment]
@jmencak Any updates here? Our team is looking to get all 4.0 docs completed for a total QE review by February 22.

**Comment by jmencak** [comment]
Yes, I think it is quite clear now the operator will not be deployed by OLM, at least for for 4.0.  It will be deployed by CVO.

**Comment by jmencak** [comment]
@ahardin-rh , I've added the "Supported tuned daemon plugins" section upstream, which should go in, especially the note about plugins *not supported*.  Also, since the operator ships in 4.0 deployed with CVO, I don't believe you need to include the "Deploying the node tuning Operator" section or it should be re-written to reflect the fact it is deployed in 4.0 automatically.  Finally, in the "Custom tuning specification", I believe it should be made clear that this is a technology preview (at least the custom profiles).  Please ping me if any questions.

/cc @jeremyeder

**Comment by ahardin-rh** [comment]
@jmencak Thank you! I made some the updates. Please review and let me know if I missed anything.

**Comment by ahardin-rh** [comment]
@openshift/team-documentation I think we're ready for a peer review. Thanks!

File: modules/custom-tuning-specification.adoc
Comment (jboxman) [inline modules/custom-tuning-specification.adoc]: logical and Operator?

File: modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc
Comment (mburke5678) [inline modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc]: Why the odd capitalization? _node tuning Operator_ looks odd.

File: modules/custom-tuning-specification.adoc
Comment (mburke5678) [inline modules/custom-tuning-specification.adoc]: The _custom resource_ the object, I think. Thus, should this be _The custom resource definition for the operator..._ ??

File: modules/custom-tuning-specification.adoc
Comment (mburke5678) [inline modules/custom-tuning-specification.adoc]: s/CR/CRD ??

File: modules/node-tuning-operator-supported-tuned-daemon-plug-ins.adoc
Comment (mburke5678) [inline modules/node-tuning-operator-supported-tuned-daemon-plug-ins.adoc]: Can these two sentences, line 27 and 30, be combined?

File: modules/node-tuning-operator.adoc
Comment (mburke5678) [inline modules/node-tuning-operator.adoc]: s/custom tuning  specification/the custom tuning  specification ?

File: modules/node-tuning-operator.adoc
Comment (mburke5678) [inline modules/node-tuning-operator.adoc]: Does the user know what a _containerized tuned daemon_ is?

File: modules/custom-tuning-specification.adoc
Comment (mburke5678) [inline modules/custom-tuning-specification.adoc]: The previous paragraphs confuse me; they seem to switch back and forth between matching and not matching. Is there a flow so that _If they match, then..._ _If they don't match, then..._

File: modules/node-tuning-operator.adoc
Comment (mburke5678) [inline modules/node-tuning-operator.adoc]: Would the majority of user know what this feature does? What is the user trying to do? Is the point to match profiles or prevent non-matched profiles?

**Comment by mburke5678** [comment]
@ahardin-rh A few questions and nits.

File: modules/custom-tuning-specification.adoc
Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: This was discussed in the upstream PR, where I have the _and_ and _or_ logical **o**perators in italics and logical **o**perators with lowercase **o**.  See https://en.wikipedia.org/wiki/Logical_connective

Perhaps just use the same formatting as the `true`/`false` values?  Leaving up to the technical writers, but needs to be consistent in the whole document.

File: modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc
Comment (jmencak) [reply modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc]: Perhaps this is an attempt to distinguish the node-tuning Operator itself vs. the logical operators?  I don't mind having the node-tuning operator lowercase, but it needs to be consistent across the OCP documentation with other sections that talk about [Oo]perators.

File: modules/custom-tuning-specification.adoc
Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: I guess I'm fine with both, but the point was to have (CR) included here so we can refer to it later.

File: modules/custom-tuning-specification.adoc
Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: Why? https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/
You're creating a CRD's instance, a CR.

File: modules/node-tuning-operator.adoc
Comment (jmencak) [reply modules/node-tuning-operator.adoc]: Does he need to?  If so, a reference to a github repo https://github.com/redhat-performance/tuned or a cross-reference to other RH docs?

File: modules/custom-tuning-specification.adoc
Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: Perhaps a decision diagram would help here?  Not sure how to create one in asciidoc.  The flow is from the highest priority (from the top in the diagram below):

    priority 10
    (pod_label: tuned.openshift.io/elasticsearch) && (node_label: node-role.kubernetes.io/master || node_label: node-role.kubernetes.io/infra) --- true ---> use profile: openshift-control-plane-es
       |
     false
       |
       V  
    priority 20
    (node_label: node-role.kubernetes.io/master || node_label: node-role.kubernetes.io/infra) --- true ---> use profile: openshift-control-plane
       |
     false
       |
       V  
    priority 30 (catch-all, always true ---> use profile: openshift-node

File: modules/node-tuning-operator.adoc
Comment (jmencak) [reply modules/node-tuning-operator.adoc]: The sentence "is automatically deployed" sounds a bit strange to me.  It is automatically deployed by the cluster version operator (CVO).  Or I'd indicate that the user doesn't have to do anything to get it deployed as it ships as part of the standard installation and the user doesn't have to do anything to have it running.

The point of the operator for 4.0 is to:
1) have a basic out-of-the box system tuning in the OCP cluster
2) make sure the applications shipped with 4.0 which have different tuning needs that the OS defaults run without issues
3) give the user a chance to specify their own tuning for their custom applications (this is a technology preview)

**Comment by jmencak** [comment]
@ahardin-rh thanks for the changes, didn't see the preview at https://node-tuning-operator--ocpdocs.netlify.com/  Links?  Thanks.

File: modules/custom-tuning-specification.adoc
Comment (jmencak) [inline modules/custom-tuning-specification.adoc]: Need to make sure _or_ is a distinguished as a logical *o*perator or just replace the logical operator with _connective_ to avoid confusion.

File: modules/custom-tuning-specification.adoc
Comment (jmencak) [inline modules/custom-tuning-specification.adoc]: Consistency, either everything `false`/`true` or just false/true, please.

File: modules/custom-tuning-specification.adoc
Comment (jmencak) [inline modules/custom-tuning-specification.adoc]: true/false, `true`/`false` consistency.

File: modules/node-tuning-operator-supported-tuned-daemon-plug-ins.adoc
Comment (jmencak) [reply modules/node-tuning-operator-supported-tuned-daemon-plug-ins.adoc]: Don't see why not. Let's combine them.

**Review by jmencak — CHANGES_REQUESTED** [review]
Thanks for the work Ashley!  Looking good, mostly consistency changes and let's make sure the operator/Operator and the logical connectives don't confuse people.  The decision diagram at the bottom would also be nice if possible.

File: modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc
Comment (kalexand-rh) [reply modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc]: At some point, we decided to always capitalize [Operator](https://github.com/openshift/openshift-docs/blob/enterprise-4.0/contributing_to_docs/term_glossary.adoc#operator).

File: modules/node-tuning-operator.adoc
Comment (kalexand-rh) [reply modules/node-tuning-operator.adoc]: If you add a link, make sure it's to docs not a [github](https://github.com/openshift/openshift-docs/blob/enterprise-4.0/contributing_to_docs/doc_guidelines.adoc#links-to-external-websites) repo.

File: modules/custom-tuning-specification.adoc
Comment (mburke5678) [reply modules/custom-tuning-specification.adoc]: I will defer to other's opinion. Perhaps my understanding isn't correct. I was looking elsewhere in our docs (1) where the file is called a CRD that creates the custom object. 
(1) https://docs.openshift.com/container-platform/3.11/admin_guide/custom_resource_definitions.html#creating-crd_admin-guide-custom-resources

Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: Unless I'm wrong:
custom resource definition (CRD) object = custom resource (CR)

The document states "the CR above is translated ...".  CRD is definitely not translated here, it is the CRD object that is translated, i.e. CR.  Therefore unless we cannot talk about CRs, we need to talk about CRD objects here.

File: modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc
Comment (adellape) [reply modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc]: And also the names of Operators. Node Tuning Operator is one of the examples listed here:

https://github.com/openshift/openshift-docs/blob/enterprise-4.0/contributing_to_docs/doc_guidelines.adoc#api-object-formatting

File: modules/custom-tuning-specification.adoc
Comment (adellape) [reply modules/custom-tuning-specification.adoc]: Would "AND logical operator" and "OR logical operator" (no quotes or markup) be too specific? IBM Style Guide seems to use that styling, at least wrt Boolean operators.

Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: I like that and I used this AND/OR in the past just like that.  Leaving up to the writers.  We also make sure people don't confuse Operator and operator/connective.

File: modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc
Comment (ahardin-rh) [reply modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc]: @adellape Yeah, I need to switch to its full proper name

File: modules/custom-tuning-specification.adoc
Comment (ahardin-rh) [reply modules/custom-tuning-specification.adoc]: @jmencak Thanks! I am working with our diagrams team to see if they can create a graphic for us.

**Comment by ahardin-rh** [comment]
@jmencak Thank you! I made some updates based on your comments. The updated preview build is ready for your review: http://file.rdu.redhat.com/~ahardin/Feb182019/node-tuning-operator/scalability_and_performance/using-node-tuning-operator.html

PTAL and let me know if I missed anything. Thanks!

File: modules/node-tuning-operator.adoc
Comment (jmencak) [inline modules/node-tuning-operator.adoc]: different tuning needs **and** default settings?  What I was trying to say originally was that if you have an OCP 4.0-shipped application that will not run with default OS (RHCoS/RHEL) settings, the operator will make sure it applies the necessary OS-level settings.  For example Elasticsearch 5.  Please see: https://github.com/openshift/cluster-node-tuning-operator/blob/master/assets/tuned/07-cr-tuned.yaml#L74 for an example.

File: modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc
Comment (jmencak) [inline modules/accessing-an-example-cluster-node-tuning-operator-specification.adoc]: I'm planning to rename `07-cr-tuned.yaml`, let's drop it and only keep `oc get Tuned/default -o yaml` as the example, that is not planned to be changed any time soon.

File: modules/custom-tuning-specification.adoc
Comment (jmencak) [inline modules/custom-tuning-specification.adoc]: Not rendered correctly and missing space.

**Comment by mburke5678** [comment]
@ahardin-rh Should I add this topic to the Nodes bucket, too?

**Comment by ahardin-rh** [comment]
@mburke5678 Sure!

**Comment by ahardin-rh** [comment]
@jmencak Thanks again for your help! I applied the latest updates and will merge this now. If we have further follow-ups, we can add them in a separate PR. I will continue working with the Brand team to create a new graphic for the decision diagram/workflow. When that is created and you have approved it, we can add it in separately as well.

**Comment by mburke5678** [comment]
Added to Nodes bucket: https://github.com/openshift/openshift-docs/pull/13669

File: modules/custom-tuning-specification.adoc
Comment (ahardin-rh) [reply modules/custom-tuning-specification.adoc]: The image is added in https://github.com/openshift/openshift-docs/pull/14056

**Pull request merged**
